plans/dht/test/find_providers.go

```diff
@@ -1,20 +1,110 @@
 package test
 
 import (
+	"context"
 	"fmt"
+	"time"
 
+	"github.com/ipfs/go-cid"
+	ipfsUtil "github.com/ipfs/go-ipfs-util"
 	"github.com/ipfs/testground/sdk/runtime"
+	"github.com/ipfs/testground/sdk/sync"
 )
 
 // FindProviders implements the Find Providers Test case
 func FindProviders(runenv *runtime.RunEnv) {
-	fmt.Printf("Not implemented yet")
+	// Test Parameters
+	var (
+		timeout     = time.Duration(runenv.IntParamD("timeout_secs", 60)) * time.Second
+		randomWalk  = runenv.BooleanParamD("random_walk", false)
+		bucketSize  = runenv.IntParamD("bucket_size", 20)
+		autoRefresh = runenv.BooleanParamD("auto_refresh", true)
+		// pProviding  = runenv.IntParamD("p_providing", 10)
+		// pResolving  = runenv.IntParamD("p_resolving", 10)
+		// pFailing    = runenv.IntParamD("p_failing", 10)
+	)
 
-	/// --- Warm Up
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	watcher, writer := sync.MustWatcherWriter(runenv)
+	defer watcher.Close()
+	defer writer.Close()
+
+
+	_, dht, _, err := SetUp(ctx, runenv, timeout, randomWalk, bucketSize, autoRefresh, watcher, writer)
+	if err != nil {
+		runenv.Abort(err)
+		return
+	}
 
 	/// --- Act I
 
+	// TODO, only `p-providing` of the nodes provide a record and store its key on redis
+
+	var (
+		cidsToPublish     []cid.Cid
+		nCidsToPublish    = 5
+		cidsToNotPublish  []cid.Cid
+		nCidsToNotPublish = 5
+	)
+
+	for i := 0; i < nCidsToPublish; i++ {
+		v := fmt.Sprintf("%d -- gonna publish", i)
```

Comment:
Log into the runenv instead?

```
runenv.Message()
```

Reply:
This string is being used only as the data that gets hashed to create the CID